base/initdefs.jl

```diff
@@ -49,31 +49,83 @@ Determine whether Julia is running an interactive session.
 """
 isinteractive() = (is_interactive::Bool)
 
+## package depots (registries, packages, environments) ##
+
+const DEPOT_PATH = String[]
+
+function init_depot_path(BINDIR = Sys.BINDIR)
+    if haskey(ENV, "JULIA_DEPOT_PATH")
+        depots = split(ENV["JULIA_DEPOT_PATH"], Sys.iswindows() ? ';' : ':')
+        push!(empty!(DEPOT_PATH), map(expanduser, depots))
+    else
+        push!(DEPOT_PATH, joinpath(homedir(), ".julia"))
+    end
+end
+
+## load-path types ##
+
+abstract type AbstractEnv end
+
+struct CurrentEnv <: AbstractEnv
+    create::Bool
+    CurrentEnv(; create::Bool=false) = new(create)
+end
+
+struct NamedEnv <: AbstractEnv
+    name::String
+    create::Bool
+    NamedEnv(name::String; create::Bool=false) = new(name, create)
+end
+
+function parse_env(env::Union{String,SubString{String}})
+    env == "@" && return CurrentEnv()
+    env == "@!" && return CurrentEnv(create=true)
+    if env[1] == '@'
+        create = env[2] == '!'
+        name = env[2+create:end]
+        name = replace(name, '#' => VERSION.major, count=1)
+        name = replace(name, '#' => VERSION.minor, count=1)
+        name = replace(name, '#' => VERSION.patch, count=1)
+        return NamedEnv(name, create=create)
+    end
+    return env # literal path
+end
+
 """
     LOAD_PATH
 
 An array of paths as strings or custom loader objects for the `require`
 function and `using` and `import` statements to consider when loading
 code.
 """
-const LOAD_PATH = String[]
+const LOAD_PATH = Any[]
 const LOAD_CACHE_PATH = String[]
 
+function parse_load_path(str::String)
+    envs = Any[split(str, Sys.iswindows() ? ';' : ':');]
+    for (i, env) in enumerate(envs)
+        if '|' in env
+            envs[i] = [parse_env(e) for e in split(env, '|')]
+        else
+            envs[i] = parse_env(env)
+        end
+    end
+    return envs
+end
+
 function init_load_path(BINDIR = Sys.BINDIR)
+    load_path = get(ENV, "JULIA_LOAD_PATH", "@|@v#.#.#|@v#.#|@v#|@default|@!v#.#")
+    append!(empty!(LOAD_PATH), parse_load_path(load_path))
     vers = "v$(VERSION.major).$(VERSION.minor)"
-    if haskey(ENV, "JULIA_LOAD_PATH")
-        prepend!(LOAD_PATH, split(ENV["JULIA_LOAD_PATH"], @static Sys.iswindows() ? ';' : ':'))
-    end
     push!(LOAD_PATH, abspath(BINDIR, "..", "local", "share", "julia", "site", vers))
     push!(LOAD_PATH, abspath(BINDIR, "..", "share", "julia", "site", vers))
```

Comment:
This currently leaves the `site` paths in here and the second one is where `stdlib` lives, which we need. However, I think that we should probably have a `Project.toml` and `Manifest.toml` for Julia itself which includes these. The installed stdlib packages should probably be in a builtin/standard depot instead, but that's kind of a detail. We can always put relative paths into `stdlib` from the manifest file.